### PR TITLE
fix increasing serive's backend counter incorrectly when restore backend

### PR DESCRIPTION
**Pull request by javu7762**
When serivce have two same backend address and port and restore backends from bpf map in lb-only mode, service's backend counter increases duplicately so backend counter can't have zero and backend naver will be deleted.

```
% cilium service update  --id=101  --frontend=3.3.3.3:80 --backends=1.1.1.1:80,1.1.1.1:80    --k8s-external
Creating new service with id '101'
Added service with 2 backends
% cilium    service list
ID    Frontend          Service Type   Backend
101   3.3.3.3:80        ExternalIPs    1 => 1.1.1.1:80
                                       2 => 1.1.1.1:80
```

So I will skip increasing backend count when backend already exits.

### CODE REVIEW

File: pkg/service/service.go [hunk 1]
```diff
@@ -1337,8 +1337,14 @@ func (s *Service) restoreServicesLocked() error {
...
 		}
 
+		backendSet := map[string]struct{}{}
 		for j, backend := range svc.Backends {
 			hash := backend.L3n4Addr.Hash()
+
+			// skip when same port and ip address exist.
+			if _, found := backendSet[hash]; found {
+				continue
+			}
```
Comment (borkmann) [inline pkg/service/service.go:1347]: Maybe I'm missing something, but who actually inserts into backendSet?

Please also add a unit test for this issue so we can assert that future changes don't break it.

**Review by borkmann — CHANGES_REQUESTED** [review]

File: pkg/service/service.go (hunk 1 above)
Comment (javu7762) [reply pkg/service/service.go:1347]: @borkmann 
I added inserts into backendSet.
and I will add a unit test for this issue.

Comment (javu7762) [reply pkg/service/service.go:1347]: I added unit test.

**Review by aditighag — COMMENTED** [review]
We already build a set while adding backends - https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L1441. Any idea why this is not being effective in filtering duplicate backend entries? AFAIK, we shouldn't have duplicate entries in the BPF maps.

**Comment by javu7762** [comment]
@aditighag 

When same backends  exist,  backend ref count can't not zero  **after pod restarts and backend is restored backends**.
So although  that backends are deleted from cmd( maybe deleted), backend exists in bpf map.

https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L1504

When  restoring backends,  backend ref count always increasing  even in same backends in one service

**Comment by aditighag** [comment]
Please see my previous comment again - https://github.com/cilium/cilium/pull/20645#pullrequestreview-1049636573. I'm asking why duplicate entries would be present in the BPF map when we already build a set while *adding* the backends? I'm not talking about the restore path, I would like to understand if/where the gap is in the backends upsert path first. Did you see duplicate entries in the BPF map, or did you notice this behavior only from code inspection?

**Comment by javu7762** [comment]
@aditighag 
Duplication entries does not exist in backend BPF map in upsert backends.
After restoring  backends, service's backend ref count is incorrect when two backends exist with same address and same port.

**Comment by aditighag** [comment]
> @aditighag Duplication entries does not exist in backend BPF map in upsert backends. After restoring backends, service's backend ref count is incorrect when two backends exist with same address and same port in one service
> when two backends exist with same address and same port in one service

When two backends with same address and same port exist where? BPF map, or somewhere else? Did you notice this from code inspection, or you hit an issue in your cluster?

**Comment by javu7762** [comment]
@aditighag 


When Standalone Load Balancer mode ( `datapath-mode: lb-only`), I can register same backend address and port twice.


```shell
#cilium service update  --id=101  --frontend=3.3.3.3:80 --backends=1.1.1.1:80,1.1.1.1:80    --k8s-external
Creating new service with id '101'
Added service with 2 backends
# cilium  service list
ID    Frontend     Service Type   Backend
101   3.3.3.3:80   ExternalIPs    1 => 1.1.1.1:80 (active)
                                  2 => 1.1.1.1:80 (active)
```

>When serivce have two same backend address and port and restore backends from bpf map in lb-only mode, service's backend counter increases duplicately so backend counter can't have zero and backend naver will be deleted.

when pod restarts and  backend ref count restores from bpf map,   same backend ref count is increasing  twice in one service.  so when deleting that service, backend node exists because ref count is not zero .

**Comment by javu7762** [comment]
I found my wrong enviroments make different results so  I close this PR.

Thanks for review and comments.

**Comment by javu7762** [comment]
@brb 
My modify sources makes wrong result so it does not be useful for others.

**Pull request closed without merging**